crates/merkle-tree/src/tree.rs

```diff
@@ -58,7 +59,8 @@ use std::{cell::RefCell, rc::Rc};
 /// A Starknet binary Merkle-Patricia tree.
 #[derive(Debug, Clone)]
 pub struct MerkleTree<H: FeltHash, const HEIGHT: usize> {
-    root: Rc<RefCell<InternalNode>>,
+    root: Option<Rc<RefCell<InternalNode>>>,
+    leaves: HashMap<BitVec<u8, Msb0>, Felt>,
```

Comment:
iiuc this is just a cache of leaves to avoid redundant traversing?

Reply:
No its not a cache. Its the leaf values that have been set so far for a given update. Since the internal nodes in the graph no longer contain the leaf values, this means the leaf values have to come from somewhere. The values for nodes from the db get queried using `Storage::leaf`, however you still need to account for the leaves which have been added but not yet stored -- since they are part of this update. These are the values stored in `leaves`.